keps/sig-node/2238-liveness-probe-grace-period/README.md

```diff
@@ -0,0 +1,491 @@
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Liveness probes currently use the `terminationGracePeriodSeconds` on both
+normal shutdown and when probes fail. Hence, if a long termination period is
+set, and a liveness probe fails, a workload will not be promptly restarted
+because it will wait for the full termination period.
+
+I propose adding a new field to liveness probes, `gracePeriodSeconds`. When
+set, it will override the `terminationGracePeriodSeconds` for liveness
+termination. This maintains the current behaviour if desired while providing
+configuration to address this unintended behaviour.
+
+## Motivation
+
+This change is important to give users the option to decouple these fields, as
+the current behaviour is counter-intuitive and doesn't provide an option to
+decouple.
+
+[Other implementation options](#alternatives) might avoid an API change, but
+would end up breaking backwards compatibility which some users may have come to
+[expect/rely on](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756100727).
+
+An example of a scenario where this bug was apparent is a long outage reported
+by an ingress controller:
+
+> the controller uses a 3600s grace period so that it can drain long requests
+> from end user app traffic (it listens on endpoints that can remain available
+> for hours like host network, but also incoming tcp load balancers which have
+> open connections), but when the controller wedges it takes an hour to
+> resolve, which is the worst possible outcome (the process wedges immediately,
+> but kube has to wait the full hour to kill it, defeating the goal of
+> liveness). [comment](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756201502)
+
+### Goals
+
+- Liveness probes can have a configurable timeout on failure.
+- Existing behaviour is not changed.
+
+### Non-Goals
+
+Any changes to probe behaviour outside of the scope of addressing the
+[bug](https://github.com/kubernetes/kubernetes/issues/64715) this is intended
+to address.
+
+## Proposal
+
+We can add a `gracePeriodSeconds` integer field to the Probe API. This field
+would specify how long the kubelet should wait before forcibly terminating the
+container when a probe fails.
+
+This change would be [API-compatible][compatible]. If the field is not
+specified, we will default to the current behaviour, which uses the
+`terminationGracePeriodSeconds`.
+
```

Comment:
Can we add some discussion here on how this value relates to initialDelaySeconds/timeoutSeconds/periodSeconds and the failureThreshold?

Also does `gracePeriodSeconds` convey the meaning of the field? (since we say "configurable timeout on failure" in the goal above)

Reply:
I am open to better names for the field.